src/ion_data/mod.rs

```diff
@@ -0,0 +1,110 @@
+///
+/// [Hash] and [Ord] are not guaranteed to be implemented for all [IonData], but when they are, they
+/// are required to be consistent with [IonEq] (and [Eq]).
+///
+/// __WARNING__—The Ion specification does _not_ define a total ordering over all Ion values. Do
+/// not depend on getting any particular result from [Ord]. Use it only as an opaque total ordering
+/// over all [IonData]. _The algorithm used for [Ord] may change between versions._
+#[derive(Debug, Clone)]
+pub struct IonData<T: IonEq>(T);
+
+impl<T: IonEq> IonData<T> {
+    /// Unwraps the value.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T: IonEq> PartialEq for IonData<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.ion_eq(&other.0)
+    }
+}
+
+impl<T: IonEq> Eq for IonData<T> {}
+
+impl<T: IonEq> From<T> for IonData<T> {
+    fn from(value: T) -> Self {
+        IonData(value)
+    }
+}
+
+impl<T: IonEq> AsRef<T> for IonData<T> {
+    fn as_ref(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<T: IonEq + Display> Display for IonData<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+impl<T: IonEq + IonOrd> PartialOrd for IonData<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(IonOrd::ion_cmp(&self.0, &other.0))
+    }
+}
+
+impl<T: IonEq + IonOrd> Ord for IonData<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        IonOrd::ion_cmp(&self.0, &other.0)
+    }
+}
+
+impl IonData<Element> {
+    /// Reads a single [IonData<Element>] from the provided data source.
+    ///
+    /// If the data source is empty, returns `Ok(None)`.
+    /// If the data source has at least one value, returns `Ok(Some(IonData<Element>))`.
+    /// If the data source has invalid data, returns `Err`.
+    pub fn read_first<A: AsRef<[u8]>>(data: A) -> IonResult<Option<IonData<Element>>> {
+        Element::read_first(data).map(|opt| opt.map(IonData))
+    }
+
+    /// Reads a single Ion [IonData<Element>] from the provided data source. If the input has invalid
+    /// data or does not contain at exactly one Ion value, returns `Err(IonError)`.
+    pub fn read_one<A: AsRef<[u8]>>(data: A) -> IonResult<IonData<Element>> {
+        Element::read_one(data).map(IonData)
+    }
+
+    /// Reads all available [IonData<Element>]s from the provided data source.
+    ///
+    /// If the input has valid data, returns `Ok(Vec<IonData<Element>>)`.
+    /// If the input has invalid data, returns `Err(IonError)`.
+    pub fn read_all<A: AsRef<[u8]>>(data: A) -> IonResult<Vec<IonData<Element>>> {
+        Element::read_all(data).map(|vec| vec.into_iter().map(IonData).collect())
+    }
+
+    delegate! {
+        to self.0 {
+            pub fn value(&self) -> &Value;
+            pub fn annotations(&self) -> &Annotations;
+        }
+    }
+}
+
+impl From<IonData<Element>> for Element {
+    fn from(value: IonData<Element>) -> Self {
+        value.0
+    }
+}
```

Comment:
I'm not sure these are worth having TBH.  Why not just provide the trivial converters and avoid the special purpose code:

```rust
impl<T> From<T> for IonData<T> where T: IonEq {
    fn from(value: T) -> Self {
        IonData<T>(value)
    }
}

impl<T> From<IonData<T>> for T where T: IonEq {
    fn from(value: IonData<T>) -> Self {
        value.0
    }
}
```

Assuming no problems with blanket conversions here, the call sites should look very reasonable:

```rust
let v: IonData<_> = Element::read_one("500").into();
```

This also works similarly for anything else that happens to not be an `Element`.

Reply:
We already have `impl<T> From<T> for IonData<T> where T: IonEq` (see ~L45).

I can't do `impl<T> From<IonData<T>> for T where T: IonEq` because:
```
error[E0210]: type parameter `T` must be covered by another type when it appears before the first local type (`IonData<T>`)
  --> src/ion_data/mod.rs:50:6
   |
50 | impl<T> From<IonData<T>> for T where T: IonEq {
   |      ^ type parameter `T` must be covered by another type when it appears before the first local type (`IonData<T>`)
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
   = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last
```

However, I did take a page out of `Cell`'s book, and provide an `into_inner()` method as part of the generic implementation of `IonData`. This implementation of `From<IonData<Element>> for Element` is redundant and could be removed.

Reply:
As for the `IonData<Element>::read_*()` functions, I think they do save you from having to repeat a lot of boilerplate elsewhere.
```rust
let elements = IonData::read_all(ion)?
// vs
let elements = Element::read_all(ion)?
            .into_iter()
            .map(IonData::from)
            .collect::<Vec<_>>()
```

...unless you can help me get a nice conversion trait impl going. I can't use `impl From<Vec<Element>> for Vec<IonData<Element>>` because of the "Only traits defined in the current crate can be implemented for arbitrary types" rule.